### PR TITLE
Add "Invisiblize" button on developer profiles (for admins)

### LINKED ISSUE #367: Add "Invisiblize" button on developer profiles (for admins)
I've had a few folks add their developer profile but fill in bogus data. "abc" as their entire bio or even "Loading...". My process is to email them and tell them I've hidden their profile then manually fire up the Rails console and set their search status to "invisible". I'd like to automate this process.

A PR addressing this issue should:
- [ ] Add a button next to the "Hire me" one (similar to #337) that says "Invisiblize"
    - [ ] Only show this button to admins
    - [ ] Use the `eye-off` icon from [Heroicons](https://heroicons.com)
- [ ] When clicked, update the `search_status` field to `:invisible` for the developer being viewed
- [ ] Send that developer an email with the below content and subject "Your profile on railsdevs was flagged"
    - [ ] Link "railsdevs" to `https://railsdevs.com`
    - [ ] Link "Your developer profile" to their profile page

Hey #{developer.name},

I'm Joe, the founder of railsdevs. Your developer profile was flagged because it doesn't contain enough information and kind of looks like spam.

I've set your profile to invisible. No one but yourself will be able to see it. You can toggle this back when you flesh out your profile a bit more.

Thanks,
Joe

### PR DESCRIPTION
**Pull request by DRBragg**
Closes #367 

In this PR we're adding the ability for the admin to flag a profile by marking it as invisible.

When the admin views a profile they feel needs to be "Invisiblized", they will click the new "Invisiblize" button marked with eye-off icon from [Heroicons](https://heroicons.com/):

invisiblize_btn

 This will trigger a POST request to `Admin::InvisiblizesController#create` and mark the developer as invisible. After successfully marking the profile as "invisible" the admin will be redirected to the developers index page with a flash notice displaying the message "Profile Invisiblized":

profile_invisiblized_message

 The `Developer` model has been updated with an `after_update_commit` callback that will only run when the `search_status` is changed to "invisible".  When this callback runs it will create a new `InvisiblizeDeveloperNotification` and email the developer that their profile was flagged:

email

You can see a [demo](#demo) video of this process below

### Requirements

- [x] Add a button next to the "Hire me" one that says "Invisiblize"
    - [x] Only show this button to admins
    - [x] Use the eye-off icon from [Heroicons](https://heroicons.com/)
- [x] When clicked, update the search_status field to :invisible for the developer being viewed
- [x] Send that developer an email with the below content and subject "Your profile on railsdevs was flagged"
    - [x] Link "railsdevs" to https://railsdevs.com
    - [x] Link "Your developer profile" to their profile page

### Demo

https://user-images.githubusercontent.com/26798548/162557768-c32509e3-7c6f-4e84-a581-44391115e572.mov

### Pull request checklist

- [x] Your code contains tests relevant for code you modified
- [x] You have linted and tested the project with `bin/check`
- [x] You have added significant changes and product updates to the [changelog](CHANGELOG.md)

### CODE REVIEW

**Comment by DRBragg** [comment]
@joemasilotti I just noticed, for some reason, I named the mailer action `flagged` while everything else in this feature references "invisiblize" or "invisible". I'm happy to change the naming of this (or anything else in the PR) if you think it'll be confusing down the line.

File: CHANGELOG.md [hunk 1]
```diff
@@ -4,6 +4,7 @@
 
 ### April
 
+* April 10 - Add "Invisiblize" button on developer profiles (for admins) #370
```
Comment (joemasilotti) [inline CHANGELOG.md:7]:
```suggestion
* April 10 - Added "Invisiblize" button on developer profiles (for admins) #370
```

File: app/controllers/admin/invisiblizes_controller.rb [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    def create
+      developer.invisible!
+      redirect_to developers_path, notice: t(".created")
+    end
+
+    private
+
+    def developer
+      @developer ||= Developer.find(params[:developer_id])
+    end
```
Comment (joemasilotti) [inline app/controllers/admin/invisiblizes_controller.rb:12]: Let's move this to line 4 as `Developer.find(params[:developer_id]).invisible!`.

File: app/models/developer.rb [hunk 3]
```diff
@@ -85,4 +86,12 @@ def missing_fields?
   def send_admin_notification
     NewDeveloperProfileNotification.with(developer: self).deliver_later(User.admin)
   end
+
+  def changed_to_invisible?
+    search_status_previously_changed? to: "invisible"
```
Comment (joemasilotti) [inline app/models/developer.rb:91]:
```suggestion
    search_status_previously_changed?(to: "invisible")
```

Also, TIL about this method!

File: app/notifications/invisiblize_developer_notification.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+  param :developer
+
+  def developer_profile_url
+    developer_path(developer)
+  end
+
+  def home_url
+    root_path
+  end
```
Comment (joemasilotti) [inline app/notifications/invisiblize_developer_notification.rb:13]: What are these two methods used for?

File: app/views/developer_mailer/flagged.html.erb [hunk 5]
```diff
@@ -0,0 +1,14 @@
+<p>Hey <%= @developer.name %>,</p>
+
+<p>
+  I'm Joe, the founder of <%= link_to "railsdevs", root_url %>. <%= link_to "Your developer profile", developer_url(@developer) %> was flagged because it doesn't contain enough information and kind of looks like spam.
+</p>
+
+<p>
+  I've set your profile to invisible. No one but yourself will be able to see it. You can toggle this back when you flesh out your profile a bit more.
```
Comment (joemasilotti) [inline app/views/developer_mailer/flagged.html.erb:8]:
```suggestion
  I've set your profile to invisible. No one but yourself is able to see it. You can toggle this back to visible after you add more details to your profile.
```

File: app/views/developer_mailer/flagged.text.erb [hunk 6]
```diff
@@ -0,0 +1,8 @@
+Hey <%= @developer.name %>,
+
+I'm Joe, the founder of railsdevs (<%= root_url %>). Your developer profile (<%= developer_url(@developer) %>) was flagged because it doesn't contain enough information and kind of looks like spam.
+
+I've set your profile to invisible. No one but yourself will be able to see it. You can toggle this back when you flesh out your profile a bit more.
```
Comment (joemasilotti) [inline app/views/developer_mailer/flagged.text.erb:5]:
```suggestion
I've set your profile to invisible. No one but yourself is able to see it. You can toggle this back to visible after you add more details to your profile.
```

File: test/mailers/previews/developer_mailer_preview.rb [hunk 7]
```diff
@@ -0,0 +1,6 @@
+class DeveloperMailerPreview < ActionMailer::Preview
+  def flagged
```
Comment (joemasilotti) [inline test/mailers/previews/developer_mailer_preview.rb:2]:
```suggestion
  def inivisiblize
```

**Review by joemasilotti — CHANGES_REQUESTED** [review]
This is looking good! I requested a few minor changes mostly related to code formatting and copy.

Outside of that, we need a better "hook" for triggering the email. In it's current state if a developer manually sets their status to invisible they will receive the notification.

File: app/models/developer.rb (hunk 3 above)
Comment (DRBragg) [reply app/models/developer.rb:91]: I forget when they changed dirty tracking (we're still on the only methods at work) but I do remember they changed it so I had to look it up too 😅

File: app/notifications/invisiblize_developer_notification.rb (hunk 4 above)
Comment (DRBragg) [reply app/notifications/invisiblize_developer_notification.rb:13]: Nothing now! Sorry about that, I had them in for something else.

File: test/mailers/previews/developer_mailer_preview.rb (hunk 7 above)
Comment (DRBragg) [reply test/mailers/previews/developer_mailer_preview.rb:2]: Agreed. I dont know why I went with `flagged` and I only caught it after I opened the PR.

**Comment by joemasilotti** [comment]
I think we can add an `invisiblize!` method to the developer that makes the change then sends the email. Instead of using a callback.

**Comment by DRBragg** [comment]
> I think we can add an `invisiblize!` method to the developer that makes the change then sends the email. Instead of using a callback.

I agree

**Comment by joemasilotti** [comment]
Looks great! Can you double check that the mailer preview works? Then I'll give it another round of tests and and merge.

**Comment by DRBragg** [comment]
> Looks great! Can you double check that the mailer preview works? Then I'll give it another round of tests and and merge.

😆 Sorry about that, didn't notice the spelling mistake first time. Should be all fixed up now.  What do you think about changing the method name in `DeveloperMailer` from `flagged` to `invisiblized`? To be honest I'm not sure what I was thinking naming it flagged

**Comment by joemasilotti** [comment]
Yeah let's go with invisibilize.

**Comment by joemasilotti** [comment]
I had a small copy edit to make so I went and renamed the method.

Thanks so much for your contribution! I'm going to merge this in as soon as CI passes.

**Pull request merged**